File: typescript/src/loader.ts
/**
 * Config loader and API for apiconf.
 */

import { readFileSync, existsSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import { parse } from "smol-toml";
import {
  AppNotFoundError,
  ConfigNotFoundError,
  KeyNotFoundError,
  ParseError,
} from "./errors.js";

/** Warning handler — replace to suppress or redirect warnings. */
export let onWarn: (message: string) => void = (msg) => console.warn(msg);
import type { AppConfig, RawConfig, RawKey } from "./types.js";

/**
 * Get the path to the config file.
 */
export function getConfigPath(): string {
  return join(homedir(), ".config", "apiconf", "config.toml");
}

/**
 * Load and parse the config file.
 */
export function loadConfig(): RawConfig {
  const path = getConfigPath();

  if (!existsSync(path)) {
    throw new ConfigNotFoundError(path);
  }

  try {
    const content = readFileSync(path, "utf-8");
    return parse(content) as RawConfig;
  } catch (error) {
    if (error instanceof ConfigNotFoundError) {
      throw error;
    }
    const message = error instanceof Error ? error.message : String(error);
    throw new ParseError(path, message);
  }
}

/**
 * Create an AppConfig from providers and keys mappings.
 */
function createAppConfig(
  appName: string,
  providers: Record<string, string>,
  keys: Record<string, RawKey>,
): AppConfig {
  const config: Record<string, string> = {};

  for (const [provider, keyName] of Object.entries(providers)) {
    const keyInfo = keys[keyName];
    if (!keyInfo) {
      onWarn(
        `Warning: Key '${keyName}' referenced by app '${appName}' not found, skipping`,
      );
      continue;
    }
    if (!keyInfo.value) {
      continue;
    }

    config[provider] = keyInfo.value;

    // Special case for ollama: also set ollama_api_base
    if (provider === "ollama") {
      config["ollama_api_base"] = keyInfo.value;
    }
  }

  return new Proxy(config, {
    get(target, prop: string | symbol) {
      if (typeof prop === "symbol" || prop in target) {
        return target[prop as string];
      }
      const available = Object.keys(target).join(", ") || "none";
      throw new Error(
        `App '${appName}' has no provider '${prop}'. Available: ${available}`,
      );
    },
  }) as AppConfig;
}

/**
 * Load configuration for an application.
 *
 * @param appName - The name of the application to load config for.
 * @returns An AppConfig object with provider key access.
 * @throws AppNotFoundError if the app is not found in the config.
 * @throws ConfigNotFoundError if the config file does not exist.
 * @throws ParseError if the config file cannot be parsed.
 */
export function load(appName: string): AppConfig {
  const config = loadConfig();

  const apps = config.apps ?? {};
  if (!(appName in apps)) {
    throw new AppNotFoundError(appName, Object.keys(apps));
  }

  const providers = apps[appName] ?? {};
  const keys = config.keys ?? {};

  return createAppConfig(appName, providers, keys);
}

/**
 * Get a key value directly by name.
 *
 * @param keyName - The name of the key to retrieve.
 * @returns The key value.
 * @throws KeyNotFoundError if the key is not found in the config.
 * @throws ConfigNotFoundError if the config file does not exist.
 * @throws ParseError if the config file cannot be parsed.
 */
export function getKey(keyName: string): string {
  const config = loadConfig();

  const keys = config.keys ?? {};
  if (!(keyName in keys)) {
    throw new KeyNotFoundError(keyName, Object.keys(keys));
  }

  const keyInfo = keys[keyName];
  if (!keyInfo || typeof keyInfo.value !== "string" || keyInfo.value === "") {
    throw new ParseError(
      getConfigPath(),
      `Key '${keyName}' is missing 'value' field`,
    );
  }

  return keyInfo.value;
}

File: python/src/apiconf/loader.py
"""Config loader and API for apiconf."""

import tomllib
from pathlib import Path
from typing import Any

from .exceptions import AppNotFoundError, ConfigNotFoundError, KeyNotFoundError, ParseError


def get_config_path() -> Path:
    """Get the path to the config file."""
    return Path.home() / ".config" / "apiconf" / "config.toml"


def load_config() -> dict[str, Any]:
    """Load and parse the config file."""
    path = get_config_path()

    if not path.exists():
        raise ConfigNotFoundError(path)

    try:
        with path.open("rb") as f:
            return tomllib.load(f)
    except tomllib.TOMLDecodeError as e:
        raise ParseError(path, str(e)) from e


class AppConfig:
    """Configuration for an application with provider key access."""

    def __init__(
        self, app_name: str, providers: dict[str, str], keys: dict[str, dict[str, str]]
    ) -> None:
        self._app_name = app_name
        self._providers = providers
        self._keys = keys
        self._resolved: dict[str, str] = {}

        # Resolve all provider -> key -> value mappings
        for provider, key_name in providers.items():
            if key_name not in keys:
                import warnings

                warnings.warn(
                    f"Key '{key_name}' referenced by app '{app_name}' not found, skipping",
                    stacklevel=2,
                )
                continue

            key_info = keys[key_name]
            if "value" not in key_info:
                continue

            self._resolved[provider] = key_info["value"]

    def __getattr__(self, name: str) -> str:
        if name.startswith("_"):
            raise AttributeError(f"'{type(self).__name__}' has no attribute '{name}'")

        if name in self._resolved:
            return self._resolved[name]

        raise AttributeError(
            f"App '{self._app_name}' has no provider '{name}'. "
            f"Available: {', '.join(self._resolved.keys()) or 'none'}"
        )

    def __repr__(self) -> str:
        return f"AppConfig({self._app_name!r}, providers={list(self._resolved.keys())})"


def load(app_name: str) -> AppConfig:
    """Load configuration for an application.

    Args:
        app_name: The name of the application to load config for.

    Returns:
        An AppConfig object with provider key access.

    Raises:
        AppNotFoundError: If the app is not found in the config.
        ConfigNotFoundError: If the config file does not exist.
        ParseError: If the config file cannot be parsed.
    """
    config = load_config()

    apps = config.get("apps", {})
    if app_name not in apps:
        raise AppNotFoundError(app_name, list(apps.keys()))

    providers = apps[app_name]
    keys = config.get("keys", {})

    return AppConfig(app_name, providers, keys)


def get_key(key_name: str) -> str:
    """Get a key value directly by name.

    Args:
        key_name: The name of the key to retrieve.

    Returns:
        The key value.

    Raises:
        KeyNotFoundError: If the key is not found in the config.
        ConfigNotFoundError: If the config file does not exist.
        ParseError: If the config file cannot be parsed.
    """
    config = load_config()

    keys = config.get("keys", {})
    if key_name not in keys:
        raise KeyNotFoundError(key_name, list(keys.keys()))

    key_entry = keys[key_name]
    if "value" not in key_entry:
        raise ParseError(get_config_path(), f"Key '{key_name}' is missing 'value' field")

    return key_entry["value"]
